src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/InstructionEncoder.cs

```diff
@@ -401,23 +418,47 @@ public LabelHandle DefineLabel()
         public void Branch(ILOpCode code, LabelHandle label)
         {
             // throws if code is not a branch:
-            int size = code.GetBranchOperandSize();
+            int operandSize = code.GetBranchOperandSize();
+            // We want the offset before we add the opcode.
+            int ilOffset = Offset;
 
-            GetBranchBuilder().AddBranch(Offset, label, code);
             OpCode(code);
+            LabelOperand(code, label, operandSize, ilOffset);
+        }
 
-            // -1 points in the middle of the branch instruction and is thus invalid.
-            // We want to produce invalid IL so that if the caller doesn't patch the branches
-            // the branch instructions will be invalid in an obvious way.
-            if (size == 1)
-            {
-                CodeBuilder.WriteSByte(-1);
-            }
-            else
+        /// <summary>
+        /// Starts encoding a switch instruction.
+        /// </summary>
+        /// <param name="branchCount">The number of branches the instruction will have.</param>
+        /// <returns>A <see cref="SwitchInstructionEncoder"/> that will
+        /// be used to emit the labels for the branches.</returns>
+        /// <remarks>
+        /// Before using this <see cref="InstructionEncoder"/> in any other way,
+        /// the method <see cref="SwitchInstructionEncoder.Branch(LabelHandle)"/>
+        /// must be called on the returned value exactly <paramref name="branchCount"/>
+        /// times. Failure to do so will result in silent bad IL code generation.
```

Comment:
Is it possible to throw instead of silently writing bad IL?

Reply:
We typically don't validate that correct IL gets emitted. I can make it throw though, if too many or too few switch branches have been emitted.

Reply:
Done. Let me know what you think.

Reply:
> Done. Let me know what you think.

Thanks, looks good

> We typically don't validate that correct IL gets emitted.

@tmat @steveharter what do you think about adding such validation?

Reply:
We don't normally validate IL including verifying even the operands for a particular opcode, so this is fine IMO.

There have been requests to do more validation, however including [this one](https://github.com/dotnet/runtime/issues/40620) by @teo-tsirpanis to at least do some basic checks.

Reply:
So I should keep d5c7793 @steveharter?